Benchmark Chainstate::ConnectBlock duration

Eunovo opened the pull request:
Introduce benchmarks to evaluate ConnectBlock performance for:
- Blocks containing only Schnorr signatures
- Blocks containing both Schnorr and ECDSA signatures
- Blocks containing only ECDSA signatures

The benchmarks in this PR focus on signature validation. Additional benchmarks may be added in the future to assess other aspects of ConnectBlock.

This is the first step toward implementing Batch Verification of Schnorr Signatures in Core. It provides a way to test and measure the performance improvements of batch verification on Core.
For more details on batch validation, refer to the [batch-verify module on secp](https://github.com/bitcoin-core/secp256k1/pull/1134) and [batch-verify on core](https://github.com/bitcoin/bitcoin/pull/29491).

Eunovo committed:
Benchmark Chainstate::ConnectBlock duration

Measure ConnectBlock performance for
- blocks containing only schnorr sigs
- blocks containing both schnorr and ecdsa sigs
- blocks containing only ecdsa sigs

This will allow testing and measurement of performance improvement for features like
batch verification of schnorr signatures

l0rinc commented on src/bench/check_connectblock.cpp:
```diff
@@ -0,0 +1,93 @@
+
+    // Create a block with (500*4) schnorr signatures to verify
+    // Typical blocks have high number of txs with few inputs
+    const int num_txs = 500;
+    const int num_inputs = 4;
```
Could we use an existing block from Mainnet () - to be sure we're measuring a representative case and not one skewed towards our preferences?
One with many schnorr sigs - but since we don't yet have one that only contains them, I don't think it makes sense to measure that (like we do here).

l0rinc replied:
They're always skewed, but if we use actual blocks as basis, at least we can't trick ourselves so easily.

fjahr replied:
> They're always skewed, 

Not if we use randomization.

> but if we use actual blocks as basis, at least we can't trick ourselves so easily.

We are just tricking ourselves differently.

After a brief discussion with @Eunovo and @josibake elsewhere I think there should be 3 separate benchmarks: No Schnorr Blocks, Random-mix Blocks and only schnorr blocks. This allows to see the impact of changes on each of these scenarios and the random mix blocks should cover what we currently see on chain.

l0rinc replied:
> They're always skewed - Not if we use randomization.

We always introduce bias, it's unavoidable. Let's try to come up with something that minimizes our own preference of how we wish the blocks looked like (as in this PR where the block only contained Schnorrs). We can avoid that by checking the actual usages and relying on them.

> I think there should be 3 separate benchmarks: No Schnorr Blocks, Random-mix Blocks and only schnorr blocks. 

This will likely minimize the biases that we introduce - especially, as mentioned before, if we rely on existing blocks, instead of an artificial setup that are skewed towards the settings we desire to measure.

Eunovo replied:
I've updated the PR to use 3 blocks
- all schnorr
- 50% schnorr, 50% ecdsa
- all ecdsa

If we used one block from Mainnet here, we would optimize for that block. This setup of 3 blocks is more balanced. We can always do an actual IBD benchmark or use tools like [Benchcoin](https://github.com/bitcoin-dev-tools/benchcoin) cc @josibake 

Connecting random blocks from Mainnet is tricky. At best, I would only be able to use the Mainnet block as a template to create the blocks for the benchmark. I'm not sure it's worth it to do this.

l0rinc replied:
> we would optimize for that block

So what are we optimizing for now?

fjahr commented:
Concept ACK

What's keeping this in draft status?

Eunovo commented:
@fjahr  I'm trying to use some real Mainnet blocks here instead. Block 861848 for example, has a lot of taproot inputs and could be good for measuring batch-validation impact on a taproot-heavy block.
cc @l0rinc

l0rinc commented on src/bench/check_connectblock.cpp:
```diff
@@ -0,0 +1,180 @@
+{
+    const auto test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+
+    size_t num_taproot{2};
+    size_t num_nontaproot{2};
```
As mentioned earlier, I would prefer basing the constants on data derived from actual usage. The current approach feels arbitrary and directly impacts the outcome.

fjahr replied:
Picking a random block is just as arbitrary and it would make sense to base the numbers on an outlook for the future rather than looking backwards. I think 50/50 is good as a starting point. I'm sure there has been a block in past that had exactly 50% schnorr sigs, so this argument about using a specific block is pointless. If you want to suggest a different split based on the past I would suggest taking a larger range of blocks and deriving a very specific split from that. Like over the last year it's probably going to be something around 40%. I would still prefer to use 50/50 because it's more representative of what we expect the future to be like but maybe you can convince other reviewers to use this.

l0rinc replied:
> Picking a random block is just as arbitrary

Absolutely, nobody recommended that.

> I'm sure there has been a block in past that had exactly 50% schnorr sigs

My point is that I'd appreciate putting in the work and actually testing that assumption. And I'm not recommending getting any block, but create one that is based on measured values, not guessed ones.

> Like over the last year it's probably going to be something around 40%.

Great, can we back that by actual measurements?

> it's more representative of what we expect the future to be like

How so? Whenever that changes we can easily adjust the benchmarks. Let's base it on data instead of feelings.

What I'm proposing is to take inspiration from the actual usages instead of coming up with these trivial values - only to discover later that they're not representative and don't cause any measurable change.

fjahr replied:
> Absolutely, nobody recommended that.

You literally [suggested it one day ago](https://github.com/bitcoin/bitcoin/pull/31689#discussion_r1922316363) and when @Eunovo said [he was going to follow through with it](https://github.com/bitcoin/bitcoin/pull/31689#issuecomment-2604687152) you gave him a thumbs up. That's what started this whole argument.

> Great, can we back that by actual measurements?

As I said, I think 50/50 is fine so I won't do those measurements. Unless @Eunovo wants to do them it's probably better if you do them and suggest to use them here.

> How so? Whenever that changes we can easily adjust the benchmarks. Let's base it on data instead of feelings.
>
> What I'm proposing is to take inspiration from the actual usages instead of coming up with these trivial values - only to discover later that they're not representative and don't cause any measurable speedup in reality.

These aren't feelings, they are assumptions based on what is being built in the bitcoin ecosystem. It seems more likely than not that we will see more taproot adoption rather than less.

l0rinc replied:
> You literally https://github.com/bitcoin/bitcoin/pull/31689#discussion_r1922316363

Read it again, there was no randomness suggested, rather the opposite, a well thought out investigation.

> likely than not that we will see more taproot adoption rather than less

Exactly, so why 50/50 and not 80/20?

fjahr replied:
> Exactly, so why 50/50 and not 80/20?

Why 80/20?

l0rinc replied:
Exactly!

fjahr replied:
> Read it again, there was no randomness suggested

What investigation? You suggested to pick a random block from mainnet and @Eunovo followed up with this saying "Block 861848 for example".

l0rinc replied:
Please read what I wrote, there was no random block suggestion.
All I'm saying is let's back up our choices by data instead of random values guided by feelings.
Did not expect that to be a controversial statement...

fjahr replied:
> Please read what I wrote, there was no random block suggestion.

Of course and you confirmed it with the engagement in the comment.

> All I'm saying is let's back up our choices by data instead of random values guided by feelings.

What data? Your suggestion is as arbitrary as mine as long as you have not made an actual suggestion of which data you want to use and put in the work to run the analysis. When have done that I can degrade that effort just like you degrade my comments here by calling them "feelings" because you will have to make decisions, too, like which blocks you actually include.

pinheadmz replied:
Ok guys, cool it please. I recommend a 24 hour break on this thread, or take it off github.

mzumsande commented:
> I'm trying to use some real Mainnet blocks here instead.

How would that be feasible if we don't have the UTXO set that this block is validated on?

l0rinc commented:
A similar attempt I had (based on your previous comment): https://github.com/bitcoin/bitcoin/pull/31699/commits/d90a0b8c90cd662bd2588e7df5fa6f641eebe3ba

But we can of course just copy the structure to make sure the benchmarks measure something real and not something completely made up (happens all the time with benchmarks and tests).
There is no perfect benchmark, but I'd like to investigate if we can do better than what we have currently.

Eunovo commented:
> How would that be feasible if we don't have the UTXO set that this block is validated on?

@mzumsande Copying the block structure could be feasible

Eunovo commented:
I used a script to count Tx inputs in the 848000 to 868000 block range. The non-taproot to taproot ratio is 80% to 20%. I used that to create the mixed block. The mixed block can used to gauge the effects of improvements like batch-validation on current blocks while we also see what it's effect on all schnorr taproot blocks look like.

l0rinc reviewed (changes requested):
I have started reviewing but it contains too many misunderstanding, a few quick notes:
* use `std::shared_ptr` to clarify ownership instead of get
* don't clear the vectors, recreate them, they're faster and more obvious
* use brace inits where possible
* for primitive parameters add comments on use site
* `use_schnorr` seems redundant, `num_taproot` already defines it
* `bench.unit("block").run([&] {` unit should likely be something that changes, e.g. total signatures or something that we can compare between the different runs
* use `emplace_back` to construct objects in-place
* destructure tuple returns to avoid calling `.first`
* num_txs can be a param with default value

Stopped reviewing here, will continue based on follow ups

fjahr commented:
> The non-taproot to taproot ratio is 80% to 20%. I used that to create the mixed block.

I still think that 50/50 would have been a better choice. Just from looking at the UTXO set we know that this number will be different in the future. Currently P2TR outputs take up 32% of the UTXO set. P2TR grows much faster than all other output types. The other outputs types have been mostly stagnating the past ~year (while P2TR utxos doubled in the last year, P2WPKH grew by 10%, the others didn't grow). And you can subtract provably unspendables and lost keys from the rest of the output types in much higher numbers which will lead to an even higher share of P2TR inputs in future blocks. Could all current P2TR key holder throw them away and never spend their outputs? Sure, but much less likely than that the past repeats itself.

And future blocks are much more interesting for this benchmark than past blocks because we can do all kinds of tricks and optimizations to speed up IBD but having a new tip propagate through the network as fast as possible is a much harder task.

This is my last comment on the topic of the mixed block split and I will just focus on the code from now on. I just want to reiterate that it makes a lot more sense to optimize for future blocks rather than past blocks.

ACK to 80/20 split. Please leave it and let's move on.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/31689#pullrequestreview-2569617843

> use std::shared_ptr to clarify ownership instead of get

I don't think this applies here, or is good advice in general. Shared pointers only seem useful when the object receiving a pointer possibly outlives the original, which I don't think is the case here anywhere. I'd instead suggest to take a reference of the underlying object where possible.

<details>
<summary>git diff on d54722be41246e89508f451bf6f9e27e7bb8d5f9</summary>

```diff
diff --git a/src/bench/connectblock.cpp b/src/bench/connectblock.cpp
index 51471aa3d5..ec28646b0f 100644
--- a/src/bench/connectblock.cpp
+++ b/src/bench/connectblock.cpp
@@ -18 +18 @@ CBlock CreateTestBlock(
-    std::shared_ptr<TestChain100Setup> test_setup,
+    TestChain100Setup& test_setup,
@@ -23 +23 @@ CBlock CreateTestBlock(
-    Chainstate& chainstate{test_setup->m_node.chainman->ActiveChainstate()};
+    Chainstate& chainstate{test_setup.m_node.chainman->ActiveChainstate()};
@@ -32 +32 @@ CBlock CreateTestBlock(
-    const WitnessV1Taproot coinbase_taproot{XOnlyPubKey(test_setup->coinbaseKey.GetPubKey())};
+    const WitnessV1Taproot coinbase_taproot{XOnlyPubKey(test_setup.coinbaseKey.GetPubKey())};
@@ -36,2 +36,2 @@ CBlock CreateTestBlock(
-    auto coinbase_to_spend{test_setup->m_coinbase_txns[0]};
-    const auto [first_tx, _]{test_setup->CreateValidTransaction(
+    auto coinbase_to_spend{test_setup.m_coinbase_txns[0]};
+    const auto [first_tx, _]{test_setup.CreateValidTransaction(
@@ -42 +42 @@ CBlock CreateTestBlock(
-    test_setup->CreateAndProcessBlock(std::vector{first_tx}, test_block_parent_coinbase, &chainstate);
+    test_setup.CreateAndProcessBlock(std::vector{first_tx}, test_block_parent_coinbase, &chainstate);
@@ -55 +55 @@ CBlock CreateTestBlock(
-        const auto [taproot_tx, _]{test_setup->CreateValidTransaction(
+        const auto [taproot_tx, _]{test_setup.CreateValidTransaction(
@@ -63 +63 @@ CBlock CreateTestBlock(
-    return test_setup->CreateBlock(txs, coinbase_spk, chainstate);
+    return test_setup.CreateBlock(txs, coinbase_spk, chainstate);
@@ -89 +89 @@ std::pair<std::vector<CKey>, std::vector<CScript>> CreateKeysAndScripts(const CK
-void BenchmarkConnectBlock(benchmark::Bench& bench, std::vector<CKey>& keys, std::vector<CScript>& spks, std::shared_ptr<TestChain100Setup> test_setup)
+void BenchmarkConnectBlock(benchmark::Bench& bench, std::vector<CKey>& keys, std::vector<CScript>& spks, TestChain100Setup& test_setup)
@@ -95 +95 @@ void BenchmarkConnectBlock(benchmark::Bench& bench, std::vector<CKey>& keys, std
-    Chainstate& chainstate{test_setup->m_node.chainman->ActiveChainstate()};
+    Chainstate& chainstate{test_setup.m_node.chainman->ActiveChainstate()};
@@ -111 +111 @@ static void ConnectBlockAllSchnorr(benchmark::Bench& bench)
-    const std::shared_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
@@ -113 +113 @@ static void ConnectBlockAllSchnorr(benchmark::Bench& bench)
-    BenchmarkConnectBlock(bench, keys, spks, test_setup);
+    BenchmarkConnectBlock(bench, keys, spks, *test_setup);
@@ -118 +118 @@ static void ConnectBlockMixed(benchmark::Bench& bench)
-    const std::shared_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
@@ -121 +121 @@ static void ConnectBlockMixed(benchmark::Bench& bench)
-    BenchmarkConnectBlock(bench, keys, spks, test_setup);
+    BenchmarkConnectBlock(bench, keys, spks, *test_setup);
@@ -126 +126 @@ static void ConnectBlockNoSchnorr(benchmark::Bench& bench)
-    const std::shared_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
@@ -128 +128 @@ static void ConnectBlockNoSchnorr(benchmark::Bench& bench)
-    BenchmarkConnectBlock(bench, keys, spks, test_setup);
+    BenchmarkConnectBlock(bench, keys, spks, *test_setup);
```
</details>

TheCharlatan commented:
Approach ACK

josibake commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,133 @@
+    return test_setup->CreateBlock(txs, coinbase_spk, chainstate);
+}
+
+std::pair<std::vector<CKey>, std::vector<CScript>> CreateKeysAndScripts(const CKey& coinbaseKey, size_t num_taproot, size_t num_nontaproot)
```
I know that we generically call things "scripts" but I think it would be more clear to name this something like `CreateKeysAndSignatures`, as that better indicates what we are testing, i.e., ECDSA signatures and Schnorr signatures. I'd also recommend renaming the variables to `num_ecdsa_sigs` and `num_schnorr_sigs`, or something similar.

josibake commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,133 @@
+    bench.unit("block").run([&] {
+        LOCK(cs_main);
+        CCoinsViewCache viewNew{&chainstate.CoinsTip()};
+        assert(chainstate.ConnectBlock(test_block, test_block_state, pindex.get(), viewNew, false));
```
`fJustCheck` is already false by default, so I think this would be better as:

```suggestion
        assert(chainstate.ConnectBlock(test_block, test_block_state, pindex.get(), viewNew));
```

josibake commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,133 @@
+{
+    Chainstate& chainstate{test_setup->m_node.chainman->ActiveChainstate()};
+
+    std::vector<CTxOut> outputs;
+    // Each transaction will pay to the same outputs that will be spent in the next one
+    outputs.reserve(scriptpubkeys.size());
+    for (const auto& spk : scriptpubkeys) {
+        outputs.emplace_back(COIN, spk);
+    }
```
It seems this was moved in here to deduplicate the code (which is great!), but I found this part to be a bit confusing. We only use this `scriptpubkeys` argument to create an outputs array, so why not create the `CTxOut` vector directly when we are creating the `scriptpubkeys` vector and then pass the `CTxOut` vector in as an argument? I also think a comment explicitly mentioning we are creating an same number input, same number output transaction would be worth adding.

Eunovo replied:
> so why not create the `CTxOut` vector directly when we are creating the `scriptpubkeys` vector and then pass the `CTxOut` vector in as an argument?

I updated the PR to do that.

> I also think a comment explicitly mentioning we are creating an same number input, same number output transaction would be worth adding.

I also added a comment stating this

josibake reviewed (commented):
Concept ACK

Thanks for picking this up, @Eunovo! This is a necessary precursor for the batch validation work, but also a generally useful benchmark to have. I think the motivation for the PR could be a bit more detailed, specifically highlighting that the benchmarks we are adding now are focused on signature verification. I imagine there are other aspects of `ConnectBlock` we might want to benchmark, which can be added to this file in the future. I do think it's a good idea to keep this initial PR focused on signature validation.

I ran the benchmark locally (after doing the recommended pyperf step) and I noticed the mixed block case was taking noticeably longer than the all ECDSA/all Schnorr cases. See the output below:

```
./build/src/bench/bench_bitcoin -filter="ConnectBlock.*"
|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       66,432,404.00 |               15.05 |    0.1% |  675,467,204.00 |  165,553,635.00 |  4.080 |  15,187,165.00 |    1.7% |      0.74 | `ConnectBlockAllSchnorr`
|       83,202,399.00 |               12.02 |    0.1% |  839,153,157.00 |  207,359,165.00 |  4.047 |  19,061,066.00 |    1.7% |      0.93 | `ConnectBlockMixed`
|       66,390,509.00 |               15.06 |    0.1% |  670,872,896.00 |  165,494,875.00 |  4.054 |  15,186,590.00 |    1.7% |      0.74 | `ConnectBlockNoSchnorr`
```

This was surprising to me. I noticed that for the 80/20 split, you are doing it _within_ a transaction, e.g., two taproot inputs and two ECDSA inputs in the same transaction (as opposed to a transaction with two Schnorr spends and a transaction with two ECDSA spends). My first guess is the extra time is coming from hashing the transaction multiple times for the different signature digest algorithms? Just a guess; will need to dig into the code more to understand. I think it would be good to document some of this in the comments, explaining how we are constructing the transactions and that we are only doing the simple case of key path spends for Schnorr and using `SIGHASH_ALL/DEFAULT` in these initial benchmarks.

I'd also echo @TheCharlatan 's feedback regarding taking a reference. Still in the process of reviewing, so will follow up with more feedback on the implementation. In the meanwhile, left some nits, if you end up retouching this before I finish my review

Eunovo commented:
I've updated the code to take a reference to the underlying test_setup object following @TheCharlatan and @josibake reviews. I've also added more comments explaining how the transactions in the test block are created.

Eunovo commented:
> My first guess is the extra time is coming from hashing the transaction multiple times for the different signature digest algorithms? Just a guess; will need to dig into the code more to understand.

I'm pretty sure this is the case after looking in `interpreter.cpp`. I added a comment explaining this

josibake reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/31689/commits/1c6b886465df0f00549e7d10c3bfefd27be7f1c2

I spent some time trying to come up with a way to create a mixed block of all schnorr input transactions and all ecdsa input transactions (as opposed to transactions with both schnorr and ecdsa inputs), but I couldn't find a simple way to do it. Furthermore, it felt unnecessary for what we are trying to accomplish: establish a baseline for signature validation of blocks with both schnorr and ecdsa signatures. Thought it was a bit surprising at first, I think the comment you added explaining that a mixed input transaction will take longer to validate is sufficient.

I left a few non-blocking style nits if you end up retouching, but overall this looks good!

davidgumberg commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+    test_setup.CreateAndProcessBlock(std::vector{first_tx}, test_block_parent_coinbase, &chainstate);
+
+    std::vector<CMutableTransaction> txs;
+    txs.reserve(num_txs);
```
not blocking, just a question: here and below, why `reserve()` for vectors that are outside of the benchmark's path?

Eunovo replied:
I understand that the use of `reserve` here is not critical, but, I generally use `reserve()` whenever I know the size of the vector beforehand in line with best practices

l0rinc replied:
It's good general practice and it documents the code as well, e.g:
```
    std::vector<CKey> keys{coinbaseKey};
    keys.reserve(num_taproot + num_nontaproot + 1);
````

l0rinc commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+        std::vector{coinbase_to_spend},
+        std::vector{COutPoint(coinbase_to_spend->GetHash(), 0)},
+        chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+    const auto test_block_parent_coinbase{GetScriptForDestination(coinbase_taproot)};
```
isn't `test_block_parent_coinbase` the same as `coinbase_spk`? Can we reuse it?
```C++
    const CScript coinbase_script{GetScriptForDestination(coinbase_taproot)};
```

l0rinc commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+    outputs.reserve(num_taproot + num_nontaproot);
+
+    for (size_t i{0}; i < num_nontaproot; i++) {
+        const CKey key{GenerateRandomKey()};
+        keys.emplace_back(key);
+        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV0KeyHash{key.GetPubKey()}));
```
to avoid possible copying we could take advantage of `emplace_back` constructing the key in-place:
```suggestion
        keys.emplace_back(GenerateRandomKey());
        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV0KeyHash{keys.back().GetPubKey()}));
```

l0rinc commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+
+BENCHMARK(ConnectBlockAllSchnorr, benchmark::PriorityLevel::HIGH);
+BENCHMARK(ConnectBlockMixed, benchmark::PriorityLevel::HIGH);
+BENCHMARK(ConnectBlockNoSchnorr, benchmark::PriorityLevel::HIGH);
```
for consistency:
```suggestion
BENCHMARK(ConnectBlockAllEcdsa, benchmark::PriorityLevel::HIGH);
```

l0rinc commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+}
+
+BENCHMARK(ConnectBlockAllSchnorr, benchmark::PriorityLevel::HIGH);
+BENCHMARK(ConnectBlockMixed, benchmark::PriorityLevel::HIGH);
```
These names are global, it may not be obvious what "ConnectBlockMixed" means outside of this context:
```suggestion
BENCHMARK(ConnectBlockMixedEcdsaSchnorr, benchmark::PriorityLevel::HIGH);
```

l0rinc commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+    });
+}
+
+static void ConnectBlockAllSchnorr(benchmark::Bench& bench)
```
This is what the flames look like - I haven't worked in this area yet, do you think it's representative?
<img width="1000" alt="image" src="https://github.com/user-attachments/assets/d55f64e2-749f-4b7c-b0fb-4e876f2e35af" />

and

<img width="1000" alt="image" src="https://github.com/user-attachments/assets/d0770f14-d29b-480e-ab78-8f37ab223ff1" />

l0rinc commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+    BenchmarkConnectBlock(bench, keys, outputs, *test_setup);
+}
+
+static void ConnectBlockNoSchnorr(benchmark::Bench& bench)
```
And these are the ECDSA ones:
<img width="1000" alt="image" src="https://github.com/user-attachments/assets/46c7b271-1e31-4cad-ba3d-b674acc019bb" />

and

<img width="1000" alt="image" src="https://github.com/user-attachments/assets/eb56e7fe-c76d-477f-830e-a4a679001782" />

l0rinc commented:
Forgot to post the diff with most of my suggestions (please double-check):

<details>
<summary>Suggestions</summary>

```patch
diff --git a/src/bench/connectblock.cpp b/src/bench/connectblock.cpp
--- a/src/bench/connectblock.cpp	(revision cb071470c2d9e595b5b771b153dfe64fdd1b1392)
+++ b/src/bench/connectblock.cpp	(date 1739374368802)
@@ -32,35 +32,33 @@
     const WitnessV1Taproot coinbase_taproot{XOnlyPubKey(test_setup.coinbaseKey.GetPubKey())};
 
     // Create the outputs that will be spent in the first transaction of the test block
-    // Doing this in a separate blocks excludes the validation of it's inputs from the benchmark
-    auto coinbase_to_spend{test_setup.m_coinbase_txns[0]};
+    // Doing this in a separate block excludes the validation of its inputs from the benchmark
+    const auto& coinbase_to_spend{test_setup.m_coinbase_txns[0]};
     const auto [first_tx, _]{test_setup.CreateValidTransaction(
-        std::vector{coinbase_to_spend},
-        std::vector{COutPoint(coinbase_to_spend->GetHash(), 0)},
-        chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
-    const auto test_block_parent_coinbase{GetScriptForDestination(coinbase_taproot)};
-    test_setup.CreateAndProcessBlock(std::vector{first_tx}, test_block_parent_coinbase, &chainstate);
+        {coinbase_to_spend},
+        {COutPoint(coinbase_to_spend->GetHash(), 0)},
+        chainstate.m_chain.Height() + 1, keys, outputs, {}, {})};
+    const CScript coinbase_script{GetScriptForDestination(coinbase_taproot)};
+    test_setup.CreateAndProcessBlock({first_tx}, coinbase_script, &chainstate);
 
     std::vector<CMutableTransaction> txs;
     txs.reserve(num_txs);
-    CTransactionRef input_tx{MakeTransactionRef(first_tx)};
-    for (int i{0}; i < num_txs; i++) {
+    for (int i{0}; i < num_txs; ++i) {
+        const auto& tx_to_spend{i == 0 ? first_tx : txs.back()};
+
         std::vector<COutPoint> inputs;
         inputs.reserve(outputs.size());
-
         for (size_t j{0}; j < outputs.size(); j++) {
-            inputs.emplace_back(input_tx->GetHash(), j);
+            inputs.emplace_back(tx_to_spend.GetHash(), j);
         }
 
         const auto [taproot_tx, _]{test_setup.CreateValidTransaction(
-            std::vector{input_tx}, inputs, chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+            {MakeTransactionRef(tx_to_spend)}, inputs, chainstate.m_chain.Height() + 1, keys, outputs, {}, {})};
         txs.emplace_back(taproot_tx);
-        input_tx = MakeTransactionRef(taproot_tx);
     }
 
     // Coinbase output can use any output type as it is not spent and will not change the benchmark
-    const CScript coinbase_spk{GetScriptForDestination(coinbase_taproot)};
-    return test_setup.CreateBlock(txs, coinbase_spk, chainstate);
+    return test_setup.CreateBlock(txs, coinbase_script, chainstate);
 }
 
 /*
@@ -69,53 +67,51 @@
  * - For non-Taproot outputs: Creates P2WPKH (native SegWit v0) outputs
  * - All outputs have value of 1 BTC
  */
-std::pair<std::vector<CKey>, std::vector<CTxOut>> CreateKeysAndOutputs(const CKey& coinbaseKey, size_t num_taproot, size_t num_nontaproot)
+std::pair<std::vector<CKey>, std::vector<CTxOut>> CreateKeysAndOutputs(const CKey& coinbaseKey, size_t num_schnorr, size_t num_ecdsa)
 {
     std::vector<CKey> keys{coinbaseKey};
-    keys.reserve(num_taproot + num_nontaproot + 1);
+    keys.reserve(num_schnorr + num_ecdsa + 1);
 
     std::vector<CTxOut> outputs;
-    outputs.reserve(num_taproot + num_nontaproot);
+    outputs.reserve(num_schnorr + num_ecdsa);
 
-    for (size_t i{0}; i < num_nontaproot; i++) {
-        const CKey key{GenerateRandomKey()};
-        keys.emplace_back(key);
-        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV0KeyHash{key.GetPubKey()}));
+    for (size_t i{0}; i < num_ecdsa; ++i) {
+        keys.emplace_back(GenerateRandomKey());
+        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV0KeyHash{keys.back().GetPubKey()}));
     }
 
-    for (size_t i{0}; i < num_taproot; i++) {
-        CKey key{GenerateRandomKey()};
-        keys.emplace_back(key);
-        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV1Taproot{XOnlyPubKey(key.GetPubKey())}));
+    for (size_t i{0}; i < num_schnorr; ++i) {
+        keys.emplace_back(GenerateRandomKey());
+        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV1Taproot{XOnlyPubKey(keys.back().GetPubKey())}));
     }
 
     return {keys, outputs};
 }
 
-void BenchmarkConnectBlock(benchmark::Bench& bench, std::vector<CKey>& keys, std::vector<CTxOut>& outputs, TestChain100Setup& test_setup)
+void BenchmarkConnectBlock(benchmark::Bench& bench, const std::vector<CKey>& keys, const std::vector<CTxOut>& outputs, TestChain100Setup& test_setup)
 {
     const auto test_block{CreateTestBlock(test_setup, keys, outputs)};
-    auto pindex{std::make_unique<CBlockIndex>(test_block)};
-    auto test_blockhash{std::make_unique<uint256>(test_block.GetHash())};
+    auto& chainman{test_setup.m_node.chainman};
+    auto& chainstate{chainman->ActiveChainstate()};
 
-    Chainstate& chainstate{test_setup.m_node.chainman->ActiveChainstate()};
-
-    pindex->nHeight = chainstate.m_chain.Height() + 1;
-    pindex->phashBlock = test_blockhash.get();
-    pindex->pprev = chainstate.m_chain.Tip();
+    {
+        LOCK(::cs_main);
+        auto* pindex{chainman->m_blockman.AddToBlockIndex(test_block, chainman->m_best_header)};
+        chainman->ActiveChain().SetTip(*pindex);
+    }
 
     BlockValidationState test_block_state;
     bench.unit("block").run([&] {
         LOCK(cs_main);
         CCoinsViewCache viewNew{&chainstate.CoinsTip()};
-        assert(chainstate.ConnectBlock(test_block, test_block_state, pindex.get(), viewNew));
+        assert(chainstate.ConnectBlock(test_block, test_block_state, chainman->ActiveChain().Tip(), viewNew));
     });
 }
 
 static void ConnectBlockAllSchnorr(benchmark::Bench& bench)
 {
-    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
-    auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_taproot=*/4, /*num_nontaproot=*/0)};
+    const auto test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_schnorr=*/4, /*num_ecdsa=*/0)};
     BenchmarkConnectBlock(bench, keys, outputs, *test_setup);
 }
 
@@ -123,23 +119,23 @@
  * This benchmark is expected to be slower than the AllSchnorr or NoSchnorr benchmark
  * because it uses transactions with both Schnorr and ECDSA signatures
  * which requires the transaction to be hashed multiple times for
- * the different signature allgorithms
+ * the different signature algorithms
  */
-static void ConnectBlockMixed(benchmark::Bench& bench)
+static void ConnectBlockMixedEcdsaSchnorr(benchmark::Bench& bench)
 {
-    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    const auto test_setup{MakeNoLogFileContext<TestChain100Setup>()};
     // Blocks in range 848000 to 868000 have a roughly 20 to 80 ratio of schnorr to ecdsa inputs
-    auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_taproot=*/1, /*num_nontaproot=*/4)};
+    auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_schnorr=*/1, /*num_ecdsa=*/4)};
     BenchmarkConnectBlock(bench, keys, outputs, *test_setup);
 }
 
-static void ConnectBlockNoSchnorr(benchmark::Bench& bench)
+static void ConnectBlockAllEcdsa(benchmark::Bench& bench)
 {
-    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
-    auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_taproot=*/0, /*num_nontaproot=*/4)};
+    const auto test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_schnorr=*/0, /*num_ecdsa=*/4)};
     BenchmarkConnectBlock(bench, keys, outputs, *test_setup);
 }
 
 BENCHMARK(ConnectBlockAllSchnorr, benchmark::PriorityLevel::HIGH);
-BENCHMARK(ConnectBlockMixed, benchmark::PriorityLevel::HIGH);
-BENCHMARK(ConnectBlockNoSchnorr, benchmark::PriorityLevel::HIGH);
+BENCHMARK(ConnectBlockMixedEcdsaSchnorr, benchmark::PriorityLevel::HIGH);
+BENCHMARK(ConnectBlockAllEcdsa, benchmark::PriorityLevel::HIGH);


```

</details>

davidgumberg commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+    const WitnessV1Taproot coinbase_taproot{XOnlyPubKey(test_setup.coinbaseKey.GetPubKey())};
+
+    // Create the outputs that will be spent in the first transaction of the test block
+    // Doing this in a separate blocks excludes the validation of it's inputs from the benchmark
```
As-is this makes callers that use this helper have more control over the blocks that they're testing.

It shouldn't have a terrible impact, but excluding the transaction that creates the first set of `outputs` of the size and arrangement that the caller wants to test makes these benchmark scenarios more 'ideal', since it avoids testing validation of an input whose spending conditions the caller doesn't have any control over, since `TestChain100Setup` gets to decide the coinbase outputs.

l0rinc replied:
I'm not expert in this area to decide if this is to sterile or not (i.e. testing a too narrow slice that isn't representative, where the biggest speedup doesn't cause any measurable macro change), will let others decide if this is too `micro` or not.

davidgumberg commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    std::vector<CMutableTransaction> txs;
+    txs.reserve(num_txs);
+    CTransactionRef input_tx{MakeTransactionRef(first_tx)};
```
The reason to separate the first tx is because it goes into a block of its own.  https://github.com/bitcoin/bitcoin/pull/31689/files#r1953657923

davidgumberg commented:
ACK https://github.com/bitcoin/bitcoin/pull/31689/commits/1c6b886465df0f00549e7d10c3bfefd27be7f1c2

I like that the benchmarks added here for `ConnectBlock` are idealized scenarios that map to specific paths we want to test the performance of, namely Schnorr signature validation and ECDSA signature validation. Measuring performance in realistic scenarios is valuable for establishing ground truth, but it seems to me that microbenchmarks like these are most useful when their focus is narrow, since they will never be a total substitute for a real end-to-end measurement I am willing to sacrifice a little bit of faithfulness in exchange for ease of interpretation.

Out of curiosity, I added a `P2PKH` scenario (https://github.com/davidgumberg/bitcoin/commit/b11348c50123999e413a5c1c373e108b78954375), where blocks are filled with 4 input 4 output P2PKH transactions,  and performance was very similar to `P2WPKH`:

```console
$ bench_bitcoin -filter=ConnectBlock.* -min-time=60000
```

benchmark |            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | flamegraph 
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------|:-------|
`ConnectBlockAllSchnorr` |       44,778,812.21 |               22.33 |    1.0% |  648,504,262.83 |  155,268,849.44 |  4.177 |  11,676,225.67 |    2.2% |     65.24 | [link](https://davidgumberg.github.io/fg/connectblockbench/allschnorr.html) |
`ConnectBlockMixed` |       56,068,853.07 |               17.84 |    1.1% |  808,990,254.47 |  194,613,843.82 |  4.157 |  14,939,414.67 |    2.2% |     66.10 | [link](https://davidgumberg.github.io/fg/connectblockbench/mixed.html) |
`ConnectBlockNoSchnorr` |       43,697,139.92 |               22.88 |    0.6% |  648,840,910.97 |  151,969,288.79 |  4.270 |  11,937,769.88 |    2.1% |     64.39 | [link](https://davidgumberg.github.io/fg/connectblockbench/noschnorr.html)|
`ConnectBlockPKH` |       43,607,527.65 |               22.93 |    0.3% |  652,306,790.73 |  151,637,220.73 |  4.302 |  12,033,719.03 |    2.2% |     63.64 | [link](https://davidgumberg.github.io/fg/connectblockbench/pkh.html) |

Eunovo commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+ * - For non-Taproot outputs: Creates P2WPKH (native SegWit v0) outputs
+ * - All outputs have value of 1 BTC
+ */
+std::pair<std::vector<CKey>, std::vector<CTxOut>> CreateKeysAndOutputs(const CKey& coinbaseKey, size_t num_taproot, size_t num_nontaproot)
```
I renamed `num_taproot` and `num_nontaproot` to `num_schnorr` and `num_ecdsa` respectively in https://github.com/bitcoin/bitcoin/pull/31689/commits/4b37efee90ee339b7bde7777b24a5268501cee9d

Eunovo commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+    std::vector<CTxOut> outputs;
+    outputs.reserve(num_taproot + num_nontaproot);
+
+    for (size_t i{0}; i < num_nontaproot; i++) {
```
Done in https://github.com/bitcoin/bitcoin/pull/31689/commits/4b37efee90ee339b7bde7777b24a5268501cee9d

Eunovo commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+ * This benchmark is expected to be slower than the AllSchnorr or NoSchnorr benchmark
+ * because it uses transactions with both Schnorr and ECDSA signatures
+ * which requires the transaction to be hashed multiple times for
+ * the different signature allgorithms
```
Done in https://github.com/bitcoin/bitcoin/pull/31689/commits/4b37efee90ee339b7bde7777b24a5268501cee9d

Eunovo commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+        std::vector{COutPoint(coinbase_to_spend->GetHash(), 0)},
+        chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+    const auto test_block_parent_coinbase{GetScriptForDestination(coinbase_taproot)};
+    test_setup.CreateAndProcessBlock(std::vector{first_tx}, test_block_parent_coinbase, &chainstate);
```
Done in https://github.com/bitcoin/bitcoin/pull/31689/commits/4b37efee90ee339b7bde7777b24a5268501cee9d

Eunovo commented on src/bench/connectblock.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    pindex->nHeight = chainstate.m_chain.Height() + 1;
+    pindex->phashBlock = test_blockhash.get();
+    pindex->pprev = chainstate.m_chain.Tip();
```
Adjusted the code to look more like this in https://github.com/bitcoin/bitcoin/pull/31689/commits/4b37efee90ee339b7bde7777b24a5268501cee9d

l0rinc replied:
is the` AddToBlockIndex` cheap enough to be done inside the benchmark

Eunovo commented:
Rebased https://github.com/bitcoin/bitcoin/commit/1c6b886465df0f00549e7d10c3bfefd27be7f1c2 to https://github.com/bitcoin/bitcoin/pull/31689/commits/bd27a83efcc3d678f33041ee34eeb019777a2405

I improved some variable and function names based on received feedback, and also made style changes to align with `doc/developer-notes.md`.

l0rinc reviewed (commented):
Concept ACK, please see my remaining comments:

<details>
<summary>Details</summary>

```patch
diff --git a/src/bench/connectblock.cpp b/src/bench/connectblock.cpp
--- a/src/bench/connectblock.cpp	(revision b3a64758bc09eb60858c23292ba7f1b9a41b9bb9)
+++ b/src/bench/connectblock.cpp	(date 1739451327613)
@@ -33,7 +33,7 @@
 
     // Create the outputs that will be spent in the first transaction of the test block
     // Doing this in a separate block excludes the validation of it's inputs from the benchmark
-    auto coinbase_to_spend{test_setup.m_coinbase_txns[0]};
+    const auto& coinbase_to_spend{test_setup.m_coinbase_txns[0]};
     const auto [first_tx, _]{test_setup.CreateValidTransaction(
         std::vector{coinbase_to_spend},
         std::vector{COutPoint(coinbase_to_spend->GetHash(), 0)},
@@ -90,25 +90,26 @@
     return {keys, outputs};
 }
 
-void BenchmarkConnectBlock(benchmark::Bench& bench, std::vector<CKey>& keys, std::vector<CTxOut>& outputs, TestChain100Setup& test_setup)
+void BenchmarkConnectBlock(benchmark::Bench& bench, const std::vector<CKey>& keys, const std::vector<CTxOut>& outputs, TestChain100Setup& test_setup)
 {
-    const auto test_block{CreateTestBlock(test_setup, keys, outputs)};
-    auto test_blockhash{std::make_unique<uint256>(test_block.GetHash())};
+    const auto& test_block{CreateTestBlock(test_setup, keys, outputs)};
     auto& chainman{test_setup.m_node.chainman};
     auto& chainstate{chainman->ActiveChainstate()};
 
-    BlockValidationState test_block_state;
+    LOCK(cs_main);
+    auto* pindex{chainman->m_blockman.AddToBlockIndex(test_block, chainman->m_best_header)};
+
     bench.unit("block").run([&] {
         LOCK(cs_main);
-        auto* pindex{chainman->m_blockman.AddToBlockIndex(test_block, chainman->m_best_header)};
         CCoinsViewCache viewNew{&chainstate.CoinsTip()};
+        BlockValidationState test_block_state;
         assert(chainstate.ConnectBlock(test_block, test_block_state, pindex, viewNew));
     });
 }
 
 static void ConnectBlockAllSchnorr(benchmark::Bench& bench)
 {
-    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    const auto test_setup{MakeNoLogFileContext<TestChain100Setup>()};
     auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_schnorr=*/4, /*num_ecdsa=*/0)};
     BenchmarkConnectBlock(bench, keys, outputs, *test_setup);
 }
@@ -121,7 +122,7 @@
  */
 static void ConnectBlockMixedEcdsaSchnorr(benchmark::Bench& bench)
 {
-    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    const auto test_setup{MakeNoLogFileContext<TestChain100Setup>()};
     // Blocks in range 848000 to 868000 have a roughly 20 to 80 ratio of schnorr to ecdsa inputs
     auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_schnorr=*/1, /*num_ecdsa=*/4)};
     BenchmarkConnectBlock(bench, keys, outputs, *test_setup);
@@ -129,7 +130,7 @@
 
 static void ConnectBlockAllEcdsa(benchmark::Bench& bench)
 {
-    const std::unique_ptr test_setup{MakeNoLogFileContext<TestChain100Setup>()};
+    const auto test_setup{MakeNoLogFileContext<TestChain100Setup>()};
     auto [keys, outputs]{CreateKeysAndOutputs(test_setup->coinbaseKey, /*num_schnorr=*/0, /*num_ecdsa=*/4)};
     BenchmarkConnectBlock(bench, keys, outputs, *test_setup);
 }
```

</details>

l0rinc reviewed (approved):
ACK 7edaf8b64cb2d59ada22042fee62a417e52368b8

I ran the benchmarks and tests locally.
I need a crypto expert to validate them conceptually as well, but I'm fine with the benchmarking code now.

(before merging please fix the typo in the PR description)

josibake commented:
reACK https://github.com/bitcoin/bitcoin/commit/7edaf8b64cb2d59ada22042fee62a417e52368b8

fjahr commented:
utACK 7edaf8b64cb2d59ada22042fee62a417e52368b8

ryanofsky commented:
> (before merging please fix the typo in the PR description)

Note: I edited out an extra comma in the description in case that was the typo, but PR otherwise looked ready so I merged it

monlovesmango commented:
hey all,

is it possible that `ConnectBlockMixedEcdsaSchnorr` is taking longer simply because it has a total of 5 keys/outputs as opposed to `ConnectBlockAllEcdsa` and `ConnectBlockAllSchnorr` which both have a total of 4 keys/outputs?

when I make all three have a total of 5 keys/outputs they are all comparable:

|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       60,589,406.32 |               16.50 |    0.8% |  841,408,256.25 |  208,383,527.67 |  4.038 |  18,821,289.97 |    1.7% |     64.95 | `ConnectBlockAllEcdsa`
|       61,600,777.26 |               16.23 |    0.5% |  849,087,583.51 |  211,909,775.35 |  4.007 |  18,904,391.80 |    1.7% |     64.16 | `ConnectBlockAllSchnorr`
|       61,224,757.09 |               16.33 |    0.2% |  843,700,462.04 |  210,410,191.25 |  4.010 |  19,149,296.85 |    1.7% |     63.34 | `ConnectBlockMixedEcdsaSchnorr`

same when I make all three have a total of 4 keys/outputs:
|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       49,623,666.70 |               20.15 |    0.8% |  672,342,072.96 |  169,902,638.80 |  3.957 |  15,205,308.17 |    1.7% |     64.03 | `ConnectBlockAllEcdsa`
|       49,647,852.54 |               20.14 |    0.3% |  678,295,781.02 |  170,690,886.42 |  3.974 |  15,250,284.70 |    1.7% |     64.67 | `ConnectBlockAllSchnorr`
|       49,319,385.60 |               20.28 |    0.4% |  671,059,527.55 |  169,247,546.89 |  3.965 |  15,427,653.66 |    1.7% |     64.44 | `ConnectBlockMixedEcdsaSchnorr`

note I did not use `pyperf system tune` as it didn't work on my machine (I think because I have AMD)

josibake commented:
> is it possible that `ConnectBlockMixedEcdsaSchnorr` is taking longer simply because it has a total of 5 keys/outputs as opposed to `ConnectBlockAllEcdsa` and `ConnectBlockAllSchnorr` which both have a total of 4 keys/outputs?

Great catch, I completely overlooked this when looking at the differences between the `Mixed` and `All` benchmarks! I double-checked this on my machine by running a few times (with pyperf), and got:

```
|            ns/block |             block/s |    err% |       ins/block |       cyc/block |    IPC |      bra/block |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       66,487,065.00 |               15.04 |    0.1% |  671,725,440.00 |  165,661,990.00 |  4.055 |  15,152,305.00 |    1.7% |      0.74 | `ConnectBlockAllEcdsa`
|       66,615,336.00 |               15.01 |    0.1% |  672,934,222.00 |  166,056,020.00 |  4.052 |  15,142,026.00 |    1.7% |      0.74 | `ConnectBlockAllSchnorr`
|       66,892,165.00 |               14.95 |    0.1% |  671,417,525.00 |  166,584,145.00 |  4.031 |  15,373,026.00 |    1.7% |      0.74 | `ConnectBlockMixedEcdsaSchnorr`
```

..which confirms your results. Even if there is a difference due to hashing the transaction twice in the `Mixed` case, it seems negligible. Happy to review/ACK a PR that updates this to make the benchmarks consistent.

l0rinc commented:
We've known about that inconsistency, it's why I recommended:
> bench.unit("block").run([&] { unit should likely be something that changes, e.g. total signatures or something that we can compare between the different runs

monlovesmango commented:
do you have a suggestion on what the unit should be? ConnectBlock benchmark is supposed to benchmark connecting a new block (and may be used for things other than signature benchmarking in the future) so I don't see any other unit than "block" making sense.

l0rinc commented:
But it's a different block, I don't think we can just compare the two performances

monlovesmango commented:
I have opened https://github.com/bitcoin/bitcoin/pull/32216 to make all the ConnectBlock benchmarks have a uniform 5 outputs per tx.

The main piece of context that might be good to have in that PR is whether the unequal number of outputs were there for a reason.
